de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java

```diff
@@ -0,0 +1,127 @@
+import de.fu_berlin.inf.dpp.util.NamedThreadFactory;
+
+public class NegotiationHandler implements INegotiationHandler {
+
+    private static final Logger log = Logger
+        .getLogger(NegotiationHandler.class);
+
+    private final ISarosSessionManager sessionManager;
+    private final ThreadPoolExecutor sessionExecutor = new ThreadPoolExecutor(0, 4, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(), new NamedThreadFactory(
+        "ServerNegotiationHandler-"));
+    private final ThreadPoolExecutor projectExecutor = new ThreadPoolExecutor(0, 10, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(), new NamedThreadFactory(
+        "ServerNegotiationHandler-"));
+
+
+    public NegotiationHandler(ISarosSessionManager sessionManager) {
+        sessionManager.setNegotiationHandler(this);
+        this.sessionManager = sessionManager;
+    }
+
+    @Override
+    public void handleOutgoingSessionNegotiation(
+        final OutgoingSessionNegotiation negotiation) {
+
+        sessionExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                SessionNegotiation.Status status = negotiation
+                    .start(new ConsoleProgressIndicator(System.out));
+
+                switch (status) {
+                    case OK:
+                        sessionManager.startSharingProjects(negotiation.getPeer());
+                        break;
+                    case ERROR:
+                        log.error("ERROR running session negotiation: "+negotiation.getErrorMessage());
+                        break;
+                    case REMOTE_ERROR:
+                        log.error("REMOTE_ERROR running session negotiation: "+negotiation.getErrorMessage());
+                        break;
+                    case CANCEL:
+                        log.info("Session negotiation was cancelled locally");
+                        break;
+                    case REMOTE_CANCEL:
+                        log.info("Session negotiation was cancelled by remote");
+                        break;
+                }
+            }
+        });
+    }
+
+    @Override
+    public void handleIncomingSessionNegotiation(
+        final IncomingSessionNegotiation negotiation) {
+
+        sessionExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                negotiation.localCancel("Server does not accept session invites",
+                    NegotiationTools.CancelOption.NOTIFY_PEER);
+                negotiation.accept(new NullProgressMonitor());
+            }
+        });
+    }
+
+    @Override
+    public void handleOutgoingProjectNegotiation(
+        final AbstractOutgoingProjectNegotiation negotiation) {
+
+        projectExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                ProjectNegotiation.Status status = negotiation
+                    .run(new ConsoleProgressIndicator(System.out));
```

Comment:
I believe this is debug stuff, so you should add a flag if. Normally this should be called with a null progress monitor or some logging progress monitor that logs at least beginTask and done calls.

Reply:
Do we have some rules for what counts as debug and what does not on the server?

The `ConsoleProgressIndicator` is indeed for user output and not debugging as it only works in a terminal and focuses on keeping a short backlog. Since the ProjectNegotiation can actually log quite a bit interesting information, I was assuming a user-friendly output might actually be desired.

Reply:
Are you aware that every service / daemon execution normally ends with 2&1 > somelogfile.txt or 2&1 > /dev/null

Nobody is looking at this output in a release environment. There is a reason SonarQube barks because of the System.out usage.

Reply:
Alright, this most likely looks total broken in log files, so I am now using the same flag used to check if the `ServerConsole` should be initialized.